Yuhsuan/2114 splatalogue backend query

YuHsuan-Hwang opened the pull request:
**Description**

1. Closes #2114

- Implemented Splatalogue advanced backend query based on the [document](https://docs.google.com/document/d/1xEZ4_IcyJUo5CermETjOtO96jp4oYC7fJhy7iJlB25I/edit#)
  - See "Implementation -> mode 1 and 2" in the document for the details of the query parameters
  - See "Implementation -> Headers" in the document for the details of the received data format and the corresponding headers from the current approach. I preserved the displayed columns in the table, with some minor changes:
    1. "CDMS/JPL Intensity": `0.00000` becomes empty
    ![image](https://user-images.githubusercontent.com/43841102/227103532-29a892ff-e757-418b-a3ed-6d01cdd4764c.png) --> ![image](https://user-images.githubusercontent.com/43841102/227103499-e9c4987e-a029-4f19-a027-dc9b791c528e.png)

    2. "Resolved QNs" and "Species": additional spaces
    ![image](https://user-images.githubusercontent.com/43841102/227104133-7f97e2bc-cb3a-4699-85fe-d4bc46c2dfbc.png) --> ![image](https://user-images.githubusercontent.com/43841102/227103914-de570e27-3ded-476d-aa39-69a2cc1f93db.png)

    3. almost all columns: trailing zeros are removed
    ![image](https://user-images.githubusercontent.com/43841102/227104670-ca690146-6aab-44c7-84e0-7e19b4cfc766.png) --> ![image](https://user-images.githubusercontent.com/43841102/227104421-73787615-1b60-4ae0-8a58-400d8357fff4.png)
    
    4. "Chemical Name": html tags are removed
    ![image](https://user-images.githubusercontent.com/43841102/231072419-1c32267d-599c-4c1a-967d-34de21da51b7.png) --> ![image](https://user-images.githubusercontent.com/43841102/231072281-106a81c4-8561-47ce-ba86-e3b8299136f1.png)

    5. "Species", "Resolved QNs", "Unresolved QNs": convert html entities to string
    <img width="260" alt="image" src="https://user-images.githubusercontent.com/43841102/234444880-2ec86101-0bbf-4dd8-8b43-ec4731d6346d.png"> --> <img width="260" alt="image" src="https://user-images.githubusercontent.com/43841102/234444966-50fec967-9c06-4770-8b6b-a72ed441b4f4.png">



    6. One additional line, could be due to different database versions
    ![image](https://user-images.githubusercontent.com/43841102/233928251-e7ab4e9b-2586-41a1-a87e-3130d5d6dced.png)

- Removed the connection check when opening the widget since it does not guarantee the connection is available when the users click `Query`
- Added simple unit tests, focusing on query parameters and parsing strings

2. Fixes #2119 by changing the Error object to string

ToDo:
- ~update the final url~ will be updated in the stable release

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] ~e2e test passing~ / corresponding fix added
- [x] changelog updated / ~no changelog update needed~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

YuHsuan-Hwang committed:
implement new query

YuHsuan-Hwang committed:
fix limit bug

YuHsuan-Hwang committed:
set headers

YuHsuan-Hwang committed:
store line data

YuHsuan-Hwang committed:
adjust line data

YuHsuan-Hwang committed:
set shift freq

YuHsuan-Hwang committed:
remove html tag

YuHsuan-Hwang committed:
remove old headers

YuHsuan-Hwang committed:
remove alive check; fix err alert

YuHsuan-Hwang committed:
minor change; reformat

YuHsuan-Hwang committed:
rename func; fix query param; add tests

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2114_splatalogue_backend_query

YuHsuan-Hwang committed:
reformat

YuHsuan-Hwang committed:
ignore the last data

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2114_splatalogue_backend_query

YuHsuan-Hwang committed:
avoid negative data length; fix test

YuHsuan-Hwang committed:
filter data instead of remove last line

YuHsuan-Hwang committed:
improve parsing

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2114_splatalogue_backend_query

YuHsuan-Hwang committed:
minor change

YuHsuan-Hwang committed:
fix eslint

YuHsuan-Hwang commented:
New changes:
* filter received data instead of removing the last data (which could be storing only the query string)
* improve parsing string

kswang1029 reviewed (changes requested):
@YuHsuan-Hwang in general this looks good based on some manual tests and e2e tests. A few issues spotted:
1. In the resolved QN column, there are some html tags such as <sup></sup> or <sub></sub> left, as well as others (eg <font ...>)
![Screen Shot 2023-03-28 at 11 36 17](https://user-images.githubusercontent.com/20819712/228124453-d4ce3686-06e8-4e5f-be38-b3261f1988e5.png)
![Screen Shot 2023-03-28 at 11 34 58](https://user-images.githubusercontent.com/20819712/228124467-763db7cd-6803-4800-893f-2c3a86aef7d2.png)

2. If we switch to "Center" mode to define a center and a range in mm or cm, the query process seems taking forever. This is picked up by the e2e tests.

YuHsuan-Hwang commented:
> @YuHsuan-Hwang in general this looks good based on some manual tests and e2e tests. A few issues spotted:
> 
> 1. In the resolved QN column, there are some html tags such as  or  left, as well as others (eg <font ...>)

@kswang1029 Do you mean  only the resolved QN column or also other columns?

kswang1029 commented:
> > @YuHsuan-Hwang in general this looks good based on some manual tests and e2e tests. A few issues spotted:
> > 
> > 1. In the resolved QN column, there are some html tags such as  or  left, as well as others (eg <font ...>)
> 
> @kswang1029 Do you mean only the resolved QN column or also other columns?

so far I only spotted the issue in the resolved QN column.

YuHsuan-Hwang commented:
> so far I only spotted the issue in the resolved QN column.

ok

YuHsuan-Hwang committed:
remove html tags in resolved QN column

YuHsuan-Hwang committed:
remove html tags in chemical name column

YuHsuan-Hwang committed:
compare and swap freq min and max

crocka commented on src/services/SplatalogueService.ts:
```diff
@@ -67,139 +70,136 @@ export class SplatalogueService {
-        }
+        // remove data with invalid species id and invalid species name
+        // this will remove data which only stores the query string
+        data = data.filter(line => line?.["species_id"] !== null || line?.["name"]);
```
If line is undefined, then undefined !== null would equal true. Is that what you want? Also, should the || be && instead according to your comment on line 84?

YuHsuan-Hwang replied:
> If line is undefined, then undefined !== null would equal true. Is that what you want?

You're right. I adjusted to first check if `line` is valid.

> Also, should the || be && instead according to your comment on line 84?

I would like to exclude lines with both species id and species name invalid, so I think `||` works fine.

YuHsuan-Hwang committed:
fix filtering data; add tests

YuHsuan-Hwang committed:
reformat

YuHsuan-Hwang commented:
@kswang1029 
> 1. In the resolved QN column, there are some html tags such as  or  left, as well as others (eg <font ...>)

fixed. I found html tags in chemical name column (and they exist in dev branch), so I also removed them.

![image](https://user-images.githubusercontent.com/43841102/231072419-1c32267d-599c-4c1a-967d-34de21da51b7.png) --> ![image](https://user-images.githubusercontent.com/43841102/231072281-106a81c4-8561-47ce-ba86-e3b8299136f1.png)

Another detail is that there are also additional spaces in species column, but I think this is very minor. These are updated in the description.

> 2. If we switch to "Center" mode to define a center and a range in mm or cm, the query process seems taking forever. This is picked up by the e2e tests.

fixed

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2114_splatalogue_backend_query

kswang1029 commented:
@YuHsuan-Hwang the regression spotted by the e2e test is fixed. 👍 Please convert this to formal PR and please proceed _code review_ to save time. I will perform extra manual test to know more about this new query backend meanwhile.

kswang1029 commented:
Spotted some minor parsing issues:
1. in the Species column, `&#150;` appears. It should be "-" so in the following example, the species is C6H-.
![Screen Shot 2023-04-14 at 13 32 38](https://user-images.githubusercontent.com/20819712/231950192-69ae7242-8bc1-4ea6-9d93-174b02ed45da.png)
There is also `&le;` as "<=".

2. in the Species column, Resolved QNs column, and Unresolved QNs column, possibly there are greek letters, such as `&alpha;`, `&gamma;`, `&epsilon;` (recombination lines), `&Omega;` (as state in QN). I suggest we show them as `alpha`, `epsilon`, `Omega`, etc.. so that users can still easily use the filter.
![Screen Shot 2023-04-14 at 13 42 18](https://user-images.githubusercontent.com/20819712/231951881-2c46471c-2594-49aa-9455-0549d897c418.png)

veggiesaurus reviewed (approved):
Code looks fine to me :+1:

YuHsuan-Hwang committed:
convert html entities

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2114_splatalogue_backend_query

YuHsuan-Hwang commented:
@kswang1029 I think it's not safe to convert `&xxx;` to `xxx` because there might be strings similar to `&le;`, unless we list all the possibilities.
Alternatively, I converted the symbols with `DOMParser`. Please check if this works fine, thanks.
Actually this exists in the dev branch, so I have noted it in the description.

kswang1029 commented:
> @kswang1029 I think it's not safe to convert `&xxx;` to `xxx` because there might be strings similar to `&le;`, unless we list all the possibilities.
> Alternatively, I converted the symbols with `DOMParser`. Please check if this works fine, thanks.
> Actually this exists in the dev branch, so I have noted it in the description.

yes, certainly this is a much better solution. Then shall we enable "select" in the table, so users can do copy-and-paster for filtering purposes?

kswang1029 commented:
this might work https://github.com/palantir/blueprint/blob/b0f1aff2b5a70433778e3b6661db90bdf755bfa2/packages/docs-app/src/examples/table-examples/tableSortableExample.tsx#L206

or demo https://blueprintjs.com/docs/versions/3/#table/features.sorting

(this table behaves not the way I imagined 🤦🏻‍♂️)

I think we need selection and keyboard copy. The right-click context menu is not needed.

kswang1029 commented:
> this might work https://github.com/palantir/blueprint/blob/b0f1aff2b5a70433778e3b6661db90bdf755bfa2/packages/docs-app/src/examples/table-examples/tableSortableExample.tsx#L206
> 
> or demo https://blueprintjs.com/docs/versions/3/#table/features.sorting
> 
> (this table behaves not the way I imagined 🤦🏻‍♂️)
> 
> I think we need selection and keyboard copy. The right-click context menu is not needed.

Postponing this as a future work due to its complexities released to a shared component.

kswang1029 reviewed (approved):
e2e tests are revised accordingly. No regression is observed from other e2e tests. Manual tests are fine as far as I can tell.

There is a potential performance issue: it seems with this new advanced backend query, performance degradation is observed comparing to the v3-stable spatalogue-frontend query.  This requires future investigation but it is not a showstopper of merging.

UPDATE: the speed is not stable. it might be due to internet.

YuHsuan-Hwang committed:
update changelog

YuHsuan-Hwang committed:
fix changelog typo
